src/ignoreFileHandler.ts

```diff
@@ -0,0 +1,97 @@
+import { readFileSync, existsSync } from 'fs';
+import * as path from 'path';
+import { workspace, Uri, Disposable } from 'vscode';
+import { getConfig } from './utils';
+
+const ignore = require('ignore');
+
+interface Ignorer {
+    ignores(filePath: string): boolean;
+}
+
+const nullIgnorer: Ignorer = { ignores: () => false };
+
+/**
+ * Create an ignore file handler. Will lazily read ignore files on a per-resource
+ * basis, and cache the contents until it changes.
+ */
+function ignoreFileHandler(disposables: Disposable[]) {
+    const ignorers = new Map<string, Ignorer>();
+    disposables.push({ dispose: () => ignorers.clear() });
+
+    return {
+        fileIsIgnored(filePath: string) {
+            return getIgnorerForFile(filePath).ignores(filePath);
+        },
+    };
+
+    function getIgnorerForFile(fsPath: string): Ignorer {
+        const absolutePath = getIgnorePathForFile(
+            fsPath,
+            getConfig(Uri.file(fsPath)).ignorePath
+        );
+
+        if (!absolutePath) {
+            return nullIgnorer;
+        }
+
+        if (ignorers.has(absolutePath)) {
+            return ignorers.get(absolutePath)!;
+        }
+
+        return loadIgnorer(Uri.file(absolutePath));
+    }
+
+    function loadIgnorer(ignoreUri: Uri): Ignorer {
+        let ignorer = nullIgnorer;
+
+        if (!ignorers.has(ignoreUri.fsPath)) {
+            const fileWatcher = workspace.createFileSystemWatcher(
+                ignoreUri.fsPath
+            );
+            disposables.push(fileWatcher);
+
+            fileWatcher.onDidCreate(loadIgnorer, null, disposables);
+            fileWatcher.onDidChange(loadIgnorer, null, disposables);
+            fileWatcher.onDidDelete(unloadIgnorer, null, disposables);
+        }
+        if (existsSync(ignoreUri.fsPath)) {
+            const ignoreFileContents = readFileSync(ignoreUri.fsPath, 'utf8');
+            ignorer = ignore().add(ignoreFileContents);
+        }
+
+        ignorers.set(ignoreUri.fsPath, ignorer);
+        return ignorer;
+    }
+
+    function unloadIgnorer(ignoreUri: Uri) {
+        ignorers.set(ignoreUri.fsPath, nullIgnorer);
+    }
+}
+
+function getIgnorePathForFile(
+    filePath: string,
+    ignorePath: string
+): string | null {
+    // Configuration `prettier.ignorePath` is set to `null`
+    if (!ignorePath) {
+        return null;
+    }
+    if (workspace.rootPath) {
+        return getPath(workspace.rootPath);
+    }
+    if (workspace.workspaceFolders) {
+        const folder = workspace.getWorkspaceFolder(Uri.file(filePath));
+        return folder ? getPath(folder.uri.fsPath) : null;
+    }
+
+    return null;
+
+    function getPath(relativeTo: string) {
```

Comment:
It works, but I am not a big fan of putting a function inside another function and below the `return` statement. Seems a bit odd to me. Maybe we can put it above the or below the `getIgnorePathForFile` function itself?

Also the only reason it works is because of hoisting if I'm not mistaken.

Reply:
Yup hoisting.
Could be a used once helper function.

Reply:
Are you okay with this or do you want to see it moved? Otherwise I'm going to dismiss/accept my request.

Reply:
Na it could definitely be moved at least before. It's of course nicer :-)

I still will let @azz test his own work before merging.